# Import required libraries

In [1]:
import pandas as pd
import numpy as np
from datetime import datetime


# Load the datasets

In [2]:
salesData_2017 = pd.read_csv(r"Sales Data/Sales Transactions-2017.csv")
salesData_2018 = pd.read_csv(r"Sales Data/Sales Transactions-2018.csv")
salesData_2019 = pd.read_csv(r"Sales Data/Sales Transactions-2019.csv")

In [3]:
# combine all the data into one
bronzeSalesData = pd.concat([salesData_2017,salesData_2018,salesData_2019])
bronzeSalesData.head()

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
0,1/4/2017,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00","3,380.00",NaN,"13,100.00"
1,1/4/2017,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
2,1/4/2017,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23,"11,500.00",NaN,"30,990.00"
3,1/4/2017,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
4,1/4/2017,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00","8,450.00",NaN,NaN


# Explore data for initial analysis

In [4]:
print(f"Dimensions: {bronzeSalesData.shape}")
print(f"Column Datatypes: \n{bronzeSalesData.dtypes}")
print(f"Columns:\t{bronzeSalesData.columns}")

Dimensions: (111206, 9)
Column Datatypes: 
Date              str
Voucher           str
Party             str
Product           str
Qty               str
Rate              str
Gross             str
Disc              str
Voucher Amount    str
dtype: object
Columns:	Index(['Date', 'Voucher', 'Party', 'Product', 'Qty', 'Rate', 'Gross', 'Disc',
       'Voucher Amount'],
      dtype='str')


In [5]:
print(f"Null data info:\n{bronzeSalesData.isnull().sum()}")
print(f"Statistical analysis on bronze data\n")
bronzeSalesData.describe(include='all')

Null data info:
Date               12591
Voucher            12557
Party                 40
Product            12591
Qty                12557
Rate               12558
Gross              12558
Disc              105609
Voucher Amount     83646
dtype: int64
Statistical analysis on bronze data



,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
count,98615,98649,111166,98615,98649,98648,98648,5597,27560
unique,836,10044,1994,867,546,1109,2548,371,7149
top,,,TP13,,100,,,,
freq,3053,3053,13056,3053,12528,3051,3053,3053,3053


#### Observations
The datatypes of the columns need to cleared for proper statistical information
- Date - to be updated as datetime
- Qty - to be updated as integer
- Rate, Gross, Disc, Voucher Amount - to be updated as float
- Disc - null can be updated as 0s

### Copy the raw data for initial cleaning

In [6]:
silverSalesData = bronzeSalesData.copy()

In [7]:
silverSalesData

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
0,1/4/2017,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00","3,380.00",NaN,"13,100.00"
1,1/4/2017,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
2,1/4/2017,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23,"11,500.00",NaN,"30,990.00"
3,1/4/2017,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
4,1/4/2017,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00","8,450.00",NaN,NaN
...,...,...,...,...,...,...,...,...,...
19171,10/10/2019,Sal:4935,K.SRIHARI,13*16 WHITE RK,400,16,"6,400.00",NaN,NaN
19172,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
19173,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
19174,NaN,Total,NaN,NaN,"99,284.90","175,381.65","2,203,649.50","20,680.00","2,189,014.50"


In [9]:
silverSalesData = silverSalesData.replace(r"^\s*$", np.nan, regex=True) # to convert the empty string as proper NaN values
# convert all the numeric columns into numeric from string or object datatype
numeric_cols = [
    "Qty",
    "Rate",
    "Gross",
    "Disc",
    "Voucher Amount"
]

for col in numeric_cols:
    silverSalesData[col] = (
        silverSalesData[col]
        .astype("string")
        .str.replace(",", "", regex=False)
        .str.strip()
    )

    silverSalesData[col] = pd.to_numeric(
        silverSalesData[col],
        errors="coerce"
    )

In [20]:
# convert date into datetime datatype
silverSalesData["Date"] = pd.to_datetime(
    bronzeSalesData["Date"]
        .astype("string")
        .str.strip(),
    format="%d/%m/%Y",
    errors="coerce"
)
silverSalesData.dtypes

Date              datetime64[us]
Voucher                      str
Party                        str
Product                      str
Qty                      Float64
Rate                     Float64
Gross                    Float64
Disc                     Float64
Voucher Amount           Float64
dtype: object

In [21]:
silverSalesData.head()

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
0,2017-04-01,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2.0,1690.0,3380.0,<NA>,13100.0
1,2017-04-01,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6.0,1620.0,9720.0,<NA>,<NA>
2,2017-04-01,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500.0,23.0,11500.0,<NA>,30990.0
3,2017-04-01,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6.0,1620.0,9720.0,<NA>,<NA>
4,2017-04-01,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5.0,1690.0,8450.0,<NA>,<NA>


In [22]:
for col in numeric_cols:
    source_not_empty = bronzeSalesData[col].notna()
    conversion_failed = source_not_empty & silverSalesData[col].isna()

    print(
        f"{col}: {conversion_failed.sum()} "
        "non-empty values failed numeric conversion"
    )

Qty: 3051 non-empty values failed numeric conversion
Rate: 3051 non-empty values failed numeric conversion
Gross: 3053 non-empty values failed numeric conversion
Disc: 3053 non-empty values failed numeric conversion
Voucher Amount: 3053 non-empty values failed numeric conversion


In [24]:
def normalize_source(series):
    return (
        series
        .astype("string")
        .str.strip()
        .replace("", pd.NA)
    )

In [25]:
conversion_failure_masks = {}

for col in numeric_cols:
    source_normalized = normalize_source(bronzeSalesData[col])

    failure_mask = (
        source_normalized.notna()
        & silverSalesData[col].isna()
    )

    conversion_failure_masks[col] = failure_mask

    print(
        f"{col}: {failure_mask.sum()} genuine conversion failures"
    )

Qty: 0 genuine conversion failures
Rate: 0 genuine conversion failures
Gross: 0 genuine conversion failures
Disc: 0 genuine conversion failures
Voucher Amount: 0 genuine conversion failures


In [26]:
print("Minimum date:", silverSalesData["Date"].min())
print("Maximum date:", silverSalesData["Date"].max())
print("Invalid dates:", silverSalesData["Date"].isna().sum())

Minimum date: 2017-04-01 00:00:00
Maximum date: 2019-10-10 00:00:00
Invalid dates: 15644


In [29]:
missingSummary = pd.DataFrame({
    "Missing Count": silverSalesData.isna().sum(),
    "Missing Percentage": (
        silverSalesData.isna().mean() * 100
    ).round(2)
})

missingSummary.sort_values(
    "Missing Percentage",
    ascending=False
)

,Missing Count,Missing Percentage
Disc,108662,97.71
Voucher Amount,86699,77.96
Date,15644,14.07
Product,15644,14.07
Voucher,15610,14.04
Gross,15611,14.04
Qty,15608,14.04
Rate,15609,14.04
Party,40,0.04


In [30]:
# Fill missing discount with Zero
silverSalesData["Disc"] = (
    silverSalesData["Disc"].fillna(0)
)
print(
    "Missing discount after filling:",
    silverSalesData["Disc"].isna().sum()
)

Missing discount after filling: 0


In [32]:
print(silverSalesData["Disc"].describe())
silverSalesData["Disc"].value_counts().head(10)

count       111206.0
mean       62.843803
std      4326.805435
min              0.0
25%              0.0
50%              0.0
75%              0.0
max         936348.0
Name: Disc, dtype: Float64


Disc
0.0      108662
280.0       255
130.0       182
350.0       157
100.0        99
210.0        88
50.0         77
260.0        77
390.0        71
200.0        63
Name: count, dtype: Int64

In [33]:
critical_transaction_cols = [
    "Date",
    "Voucher",
    "Party",
    "Product",
    "Qty",
    "Rate",
    "Gross"
]

transaction_mask = (
    silverSalesData[critical_transaction_cols]
    .notna()
    .all(axis=1)
)

In [34]:
print("Total rows:", len(silverSalesData))
print("Valid transaction rows:", transaction_mask.sum())
print("Other rows:", (~transaction_mask).sum())

Total rows: 111206
Valid transaction rows: 95561
Other rows: 15645


In [35]:
rejectedSalesData = (
    silverSalesData.loc[~transaction_mask]
    .copy()
)

validSalesData = (
    silverSalesData.loc[transaction_mask]
    .copy()
)
rejectedSalesData[
    [
        "Date",
        "Voucher",
        "Party",
        "Product",
        "Qty",
        "Rate",
        "Gross",
        "Disc",
        "Voucher Amount"
    ]
].head(50)

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
8,NaT,NaN,khader vali late en,NaN,<NA>,<NA>,<NA>,0.0,<NA>
9,NaT,NaN,try,NaN,<NA>,<NA>,<NA>,0.0,<NA>
11,NaT,NaN,#NAME?,NaN,<NA>,<NA>,<NA>,0.0,<NA>
12,NaT,NaN,DUMMY ENTRY,NaN,<NA>,<NA>,<NA>,0.0,<NA>
14,NaT,NaN,#NAME?,NaN,<NA>,<NA>,<NA>,0.0,<NA>
15,NaT,NaN,dummy entry,NaN,<NA>,<NA>,<NA>,0.0,<NA>
17,NaT,NaN,dummy entry,NaN,<NA>,<NA>,<NA>,0.0,<NA>
39,NaT,NaN,#NAME?,NaN,<NA>,<NA>,<NA>,0.0,<NA>
41,NaT,NaN,#NAME?,NaN,<NA>,<NA>,<NA>,0.0,<NA>
43,NaT,NaN,#NAME?,NaN,<NA>,<NA>,<NA>,0.0,<NA>


In [36]:
rejectedSalesData.isna().sum()

Date              15644
Voucher           15610
Party                40
Product           15644
Qty               15608
Rate              15609
Gross             15611
Disc                  0
Voucher Amount    15611
dtype: int64

In [38]:
def get_missing_columns(row):
    return ", ".join(
        row.index[row.isna()].tolist()
    )

rejectedSalesData["Missing Columns"] = (
    rejectedSalesData.apply(
        get_missing_columns,
        axis=1
    )
)

rejectedSalesData[
    "Missing Columns"
].value_counts().head(20)

Missing Columns
Date, Voucher, Product, Qty, Rate, Gross, Voucher Amount           15602
Date, Party, Product                                                  34
Date, Voucher, Party, Product, Qty, Rate, Gross, Voucher Amount        6
Date, Voucher, Product, Gross, Voucher Amount                          2
Rate, Gross, Voucher Amount                                            1
Name: count, dtype: int64

## Distinguish comment rows from invalid transactions
A comment or continuation row typically has no date, voucher, product, quantity, rate, or gross.

In [42]:
comment_or_continuation_mask = (
    silverSalesData["Date"].isna()
    & silverSalesData["Voucher"].isna()
    & silverSalesData["Product"].isna()
    & silverSalesData["Qty"].isna()
    & silverSalesData["Rate"].isna()
    & silverSalesData["Gross"].isna()
)

A potentially invalid transaction has identifying information but is missing a critical numeric value:

In [40]:
invalid_transaction_mask = (
    silverSalesData["Date"].notna()
    & silverSalesData["Voucher"].notna()
    & silverSalesData["Product"].notna()
    & (
        silverSalesData["Qty"].isna()
        | silverSalesData["Rate"].isna()
        | silverSalesData["Gross"].isna()
    )
)

In [43]:
print(
    "Comments or continuation rows:",
    comment_or_continuation_mask.sum()
)

print(
    "Potential invalid transactions:",
    invalid_transaction_mask.sum()
)

Comments or continuation rows: 15608
Potential invalid transactions: 1


In [44]:
silverSalesData.loc[
    invalid_transaction_mask,
    [
        "Date",
        "Voucher",
        "Party",
        "Product",
        "Qty",
        "Rate",
        "Gross",
        "Disc",
        "Voucher Amount"
    ]
].head(50)

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
43740,2018-03-05,Sal:9266,CUSTOMER,TIRML-BL-KG-17*23,2.0,<NA>,<NA>,0.0,<NA>


In [45]:
silverSalesData.loc[
    comment_or_continuation_mask,
    [
        "Date",
        "Voucher",
        "Party",
        "Product",
        "Qty",
        "Rate",
        "Gross",
        "Disc",
        "Voucher Amount"
    ]
].head(50)

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
8,NaT,NaN,khader vali late en,NaN,<NA>,<NA>,<NA>,0.0,<NA>
9,NaT,NaN,try,NaN,<NA>,<NA>,<NA>,0.0,<NA>
11,NaT,NaN,#NAME?,NaN,<NA>,<NA>,<NA>,0.0,<NA>
12,NaT,NaN,DUMMY ENTRY,NaN,<NA>,<NA>,<NA>,0.0,<NA>
14,NaT,NaN,#NAME?,NaN,<NA>,<NA>,<NA>,0.0,<NA>
15,NaT,NaN,dummy entry,NaN,<NA>,<NA>,<NA>,0.0,<NA>
17,NaT,NaN,dummy entry,NaN,<NA>,<NA>,<NA>,0.0,<NA>
39,NaT,NaN,#NAME?,NaN,<NA>,<NA>,<NA>,0.0,<NA>
41,NaT,NaN,#NAME?,NaN,<NA>,<NA>,<NA>,0.0,<NA>
43,NaT,NaN,#NAME?,NaN,<NA>,<NA>,<NA>,0.0,<NA>


### Create a row-status column
This provides an auditable cleaning process.

In [46]:
silverSalesData["Row Status"] = "Needs Review"

silverSalesData.loc[
    transaction_mask,
    "Row Status"
] = "Valid Transaction"

silverSalesData.loc[
    comment_or_continuation_mask,
    "Row Status"
] = "Comment or Continuation"

silverSalesData.loc[
    invalid_transaction_mask,
    "Row Status"
] = "Invalid Numeric Transaction"

In [47]:
silverSalesData["Row Status"].value_counts()

Row Status
Valid Transaction              95561
Comment or Continuation        15608
Needs Review                      36
Invalid Numeric Transaction        1
Name: count, dtype: int64

Then split the data:

In [48]:
rejectedSalesData = (
    silverSalesData[
        silverSalesData["Row Status"]
        != "Valid Transaction"
    ]
    .copy()
)

silverSalesData = (
    silverSalesData[
        silverSalesData["Row Status"]
        == "Valid Transaction"
    ]
    .copy()
)

silverSalesData.drop(
    columns=["Row Status"],
    inplace=True
)

In [49]:
silverSalesData.reset_index(
    drop=True,
    inplace=True
)

rejectedSalesData.reset_index(
    drop=True,
    inplace=True
)

### Clean text columns
Now standardize whitespace in the text fields.

In [50]:
text_cols = [
    "Voucher",
    "Party",
    "Product"
]

for col in text_cols:
    silverSalesData[col] = (
        silverSalesData[col]
        .astype("string")
        .str.strip()
        .str.replace(r"\s+", " ", regex=True)
    )

Preserve the original names while creating standardized versions for grouping:

Because these values might refer to the same party:
- Srinu
- SRINU
- srinu

In [51]:
silverSalesData["Party Standardized"] = (
    silverSalesData["Party"]
    .str.upper()
    .str.strip()
    .str.replace(r"\s+", " ", regex=True)
)

silverSalesData["Product Standardized"] = (
    silverSalesData["Product"]
    .str.upper()
    .str.strip()
    .str.replace(r"\s+", " ", regex=True)
)

## Analyze exact duplicate rows

In [52]:
exact_duplicate_count = (
    silverSalesData.duplicated().sum()
)

print(
    "Exact duplicate rows:",
    exact_duplicate_count
)

Exact duplicate rows: 275


In [53]:
duplicateRows = (
    silverSalesData[
        silverSalesData.duplicated(
            keep=False
        )
    ]
    .sort_values(
        by=[
            "Date",
            "Voucher",
            "Party",
            "Product"
        ]
    )
)

duplicateRows.head(50)

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount,Party Standardized,Product Standardized
1320,2017-04-12,Sal:439,TP13,Classic 250ml,100.0,23.0,2300.0,0.0,<NA>,TP13,CLASSIC 250ML
1335,2017-04-12,Sal:439,TP13,Classic 250ml,100.0,23.0,2300.0,0.0,<NA>,TP13,CLASSIC 250ML
1864,2017-04-17,Sal:443,SHAIK SIFULLA,CYCLE-BK-13*16,10.0,13.0,130.0,130.0,<NA>,SHAIK SIFULLA,CYCLE-BK-13*16
1873,2017-04-17,Sal:443,SHAIK SIFULLA,CYCLE-BK-13*16,10.0,13.0,130.0,130.0,<NA>,SHAIK SIFULLA,CYCLE-BK-13*16
1985,2017-04-17,Sal:552,TP13,Vishnu 300ml,100.0,27.5,2750.0,0.0,<NA>,TP13,VISHNU 300ML
1990,2017-04-17,Sal:552,TP13,Vishnu 300ml,100.0,27.5,2750.0,0.0,<NA>,TP13,VISHNU 300ML
2040,2017-04-18,Sal:497,Lakshmifancy,"TAPE ROLL 2"" 50MTRS",4.0,125.0,500.0,0.0,<NA>,LAKSHMIFANCY,"TAPE ROLL 2"" 50MTRS"
2041,2017-04-18,Sal:497,Lakshmifancy,"TAPE ROLL 2"" 50MTRS",4.0,125.0,500.0,0.0,<NA>,LAKSHMIFANCY,"TAPE ROLL 2"" 50MTRS"
2484,2017-04-21,Sal:599,SPM PROVISION-DAKKI,NO-00,1.0,150.0,150.0,0.0,<NA>,SPM PROVISION-DAKKI,NO-00
2485,2017-04-21,Sal:599,SPM PROVISION-DAKKI,NO-00,1.0,150.0,150.0,0.0,<NA>,SPM PROVISION-DAKKI,NO-00


In [54]:
# drop the duplicates
silverSalesData = (
    silverSalesData
    .drop_duplicates()
    .reset_index(drop=True)
)

## Validate Gross = Qty × Rate

In [60]:
silverSalesData["Calculated Gross"] = (
    silverSalesData["Qty"]
    * silverSalesData["Rate"]
)
silverSalesData["Gross Difference"] = (
    silverSalesData["Gross"]
    - silverSalesData["Calculated Gross"]
)
silverSalesData["Gross Valid"] = np.isclose(
    silverSalesData["Gross"],
    silverSalesData["Calculated Gross"],
    rtol=0,
    atol=0.01
)


In [61]:
silverSalesData[
    "Gross Valid"
].value_counts()

Gross Valid
True     95270
False       16
Name: count, dtype: int64

In [62]:
silverSalesData.loc[
    ~silverSalesData["Gross Valid"],
    [
        "Date",
        "Voucher",
        "Product",
        "Qty",
        "Rate",
        "Gross",
        "Calculated Gross",
        "Gross Difference"
    ]
].head(20)

,Date,Voucher,Product,Qty,Rate,Gross,Calculated Gross,Gross Difference
716,2017-04-07,Sal:146,CYCLE-BLU-10*12,1600.0,5.42,8670.0,8672.0,-2.0
16633,2017-08-16,Sal:4219,"TAPEROLL 2""",12.0,83.33,1000.0,999.96,0.04
19558,2017-09-08,Sal:4921,KAVITHA GOLD 10*12,1400.0,5.57,7800.0,7798.0,2.0
19613,2017-09-09,Sal:4913,KAVITHA GOLD 10*12,700.0,5.57,3900.0,3899.0,1.0
26660,2017-11-14,Sal:6592,GARBAGE BAGS-KGS,30.32,120.0,3639.0,3638.4,0.6
30083,2017-12-16,Sal:7487,DURGA 13*16 PINK,1600.0,8.94,14310.0,14304.0,6.0
33386,2018-01-17,Sal:8221,BAHUBALI WINE,110.0,22.73,2500.0,2500.3,-0.3
36138,2018-02-19,Sal:8933,KRISHNA 10*12 RED,930.0,6.21,5775.0,5775.3,-0.3
39648,2018-03-26,Sal:9841,KRISHNA 10*12 RED,1200.0,6.23,7480.0,7476.0,4.0
39649,2018-03-26,Sal:9841,KRISHNA 13*16 RED,640.0,13.06,8360.0,8358.4,1.6


## Create basic validity flags

In [63]:
silverSalesData["Invalid Qty"] = (
    silverSalesData["Qty"] <= 0
)

silverSalesData["Invalid Rate"] = (
    silverSalesData["Rate"] < 0
)

silverSalesData["Invalid Gross"] = (
    silverSalesData["Gross"] < 0
)

silverSalesData["Invalid Discount"] = (
    silverSalesData["Disc"] < 0
)

In [64]:
validationSummary = pd.Series({
    "Invalid Qty": silverSalesData[
        "Invalid Qty"
    ].sum(),

    "Invalid Rate": silverSalesData[
        "Invalid Rate"
    ].sum(),

    "Invalid Gross": silverSalesData[
        "Invalid Gross"
    ].sum(),

    "Invalid Discount": silverSalesData[
        "Invalid Discount"
    ].sum(),

    "Gross Mismatch": (
        ~silverSalesData["Gross Valid"]
    ).sum()
})

validationSummary

Invalid Qty          0
Invalid Rate         0
Invalid Gross        0
Invalid Discount     0
Gross Mismatch      16
dtype: int64

In [65]:
# Missing discount means no recorded discount
silverSalesData["Disc"] = (
    silverSalesData["Disc"].fillna(0)
)

# Identify complete transaction rows
critical_transaction_cols = [
    "Date",
    "Voucher",
    "Party",
    "Product",
    "Qty",
    "Rate",
    "Gross"
]

transaction_mask = (
    silverSalesData[critical_transaction_cols]
    .notna()
    .all(axis=1)
)

print("Total rows:", len(silverSalesData))
print("Valid transaction rows:", transaction_mask.sum())
print("Rows requiring review:", (~transaction_mask).sum())

Total rows: 95286
Valid transaction rows: 95286
Rows requiring review: 0


In [66]:
silverSalesData.loc[
    ~transaction_mask,
    [
        "Date",
        "Voucher",
        "Party",
        "Product",
        "Qty",
        "Rate",
        "Gross",
        "Disc",
        "Voucher Amount"
    ]
].head(50)

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount


## Inferences
- Main structural cleaning like numeric/date conversion, rejection of incomplete rows, text normalization, columns standardization, duplicate analysis/removal, gross validation
- The resulting dataset has 95,286 records with valid rows, while 16 gross mismatches, zero quantity etc., are present.
- The remaining business-level cleaning and recon needs to be done.